### ISSUE TITLE
XAML designer code-behinds require solution reload to make IDE aware of changes

### ISSUE THREAD

**Issue by wjk** [xamarin/Xamarin.Forms#2612]
### Description

The title says it all, really. I change the XAML in such a way that requires the designer code file be rewritten (say, add or change an `x:Name`), then I rebuild the solution. The code editor for the code-behind does _not_ update with the new value of the file until the solution is closed and reopened. Needless to say, this gets really annoying really fast.

### Steps to Reproduce

1. Download the attached test case
2. Open it in VS 2017 for Windows
3. Open MainPage.xaml, and change `x:Name="RedBox"` to read `x:Name="RedBox2"`.
4. Open the MainPage.cs file. Note how VS thinks the RedBox variable is still present, even though the compiler knows it isn't (build and you will get an error).

### Expected Behavior

Any change in the designer file should be picked up by VS automatically.

### Actual Behavior

I have to close and reload the solution to force VS to reread the generated designer files.

### Basic Information

- Version with issue: VS for Windows 15.6.7
- Last known good version: Don't know
- IDE: Visual Studio for Windows
- Platform Target Frameworks: Doesn't apply, this is a designer problem
- Android Support Library Version: N/A
- Nuget Packages: Xamarin.Forms v3.0.0-296286-pre2, no others
- Affected Devices: All of them, this is a designer problem.

### Reproduction Link

[XamFormsTest.zip](https://github.com/xamarin/Xamarin.Forms/files/1977271/XamFormsTest.zip)

**Comment by StephaneDelcroix** [comment]
a lot of intellisense changes landed in 15.7, iirc.

@kzu could you chime in ?

**Comment by kzu** [comment]
@mgoertz-msft is the right person. Is this for shared project or netstandard?

**Comment by wjk** [comment]
@kzu It's a netstandard project. Note that I haven't retested this since the official 3.0 release at BUILD, though.

**Comment by kzu** [comment]
Any chance you can try it out with just the XF nuget package and a solution with just the netstandard project? Because that's basically all (Visual Studio built-in) XAML language service + nuget, and absolutely nothing from the Xamarin tooling (the VS extension) point of view, and that should work flawlessly out of the box.

Also, are you using any other extensions, like Resharper or LiveXAML? The former completely replaces the Visual Studio XAML language service, so if that's the case, it's a bug you should report to them instead. You can turn off their XAML editing integration explicitly via an option, IIRC.

Thanks

**Comment by wjk** [comment]
I *am* just using the XF nuget package and the netstandard project. In fact, that's the sample I uploaded for my original post. Nor do I have any third-party XAML-related extensions installed. I also retested using post-BUILD Visual Studio, to no avail. This problem still occurs.

**Comment by mgoertz-msft** [comment]
@wjk Like Stephane said, you need to upgrade 15.7 to take advantage of the Xamarin Forms XAML IntelliSense improvements we have added.

However, your repro steps indicate that C# IntelliSense is incorrect when you change named XAML elements. That is an entirely different mechanism that is driven by running the Custom Tool for XAML files, i.e. the XF Markup compiler, when switching away from or saving dirty XAML files.

@kzu This mechanism seems to be broken (even in the latest 15.8 Preview). I can't even add an x:Name to an element and use it in code behind.

**Comment by kzu** [comment]
That's the underlying project system built-in behavior. We (Xamarin VS) don't do anything about it. The item basically has a Generator item metadata which is an MSBuild target, and the IDE (MSBuild(?)) should run it automatically whenever you save the file.

**Comment by mgoertz-msft** [comment]
Yes, the IDE will run whatever you specify as the Custom Tool. XF specifies "MSBuild:UpdateDesignTimeXaml" which tells VS to run msbuild with the target "UpdateDesignTimeXaml". There are multiple possible failures that could be responsible for this issue that need to be looked at:

1. VS did not run the msbuild target at all (not very likely but a possibility nevertheless).
2. There were errors building the target.
3. The CS file generated during this design-time build, assuming that this is what this target does analogous to WPF/UWP, is no longer recognized by Roslyn as part of the compilation.

**Comment by kzu** [comment]
@pauldipietro @StephaneDelcroix there's nothing special that we do in the IDE. Moreover, the project system isn't even us. So it's either something to discuss at the MSBuild level (i.e. is it a bug that should be reported to https://github.com/Microsoft/msbuild) or the NETStandard project level (i.e. is it a bug that should be reported to https://github.com/dotnet/project-system/). 

More information on how do diagnose the XF targets for this kind of issues is available at https://github.com/dotnet/project-system/blob/master/docs/design-time-builds.md too, which has helpful links to tools and a bunch of tips.
